tests/Microsoft.Bot.Builder.TestBot/ScalableDialogBotBase.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace Microsoft.Bot.Builder.TestBot
+{
+    /// <summary>
+    /// This implemnentation can be used as the basis for a scalable bot which has multiple instances running on
+    /// multiple nodes. The assumption there being the try-save operation is guarded with an eTag condition. When the
+    /// eTag condition fails the load and business logic should be repeated before the save is attempted again.
+    /// Because the BotState provider uses a shared cache the semaphore is still required in this code. As a result this
+    /// code behaves correctly but is inefficient. There are a couple of solutions: reduce the granularity of the locking
+    /// to make it specific to a conversation or use caching that is created and used inside the scope of the load/try-save.
+    /// </summary>
+    public class ScalableDialogBotBase : IBot
+    {
+        private SemaphoreSlim _semaphore;
+        private IBotStoreManager _storeManager;
+        private string _rootDialogName;
+
+        public ScalableDialogBotBase(TestBotAccessors accessors, string rootDialogName)
+        {
+            // create the DialogSet from accessor
+            Dialogs = new DialogSet(accessors.ConversationDialogState);
+
+            // a semaphore to serialize access to the bot state
+            _semaphore = accessors.SemaphoreSlim;
+
+            // the state to attempt to save at the end of every turn
+            _storeManager = accessors.StoreManager;
+
+            // the name of the root dialog this bot runs
+            _rootDialogName = rootDialogName;
+        }
+
+        protected DialogSet Dialogs { get; private set; }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            // The loop here is the classic eTag retry loop. After a save failure you need to start over
+            // with a re-load of the state. The assumption being you have implemented these abstractions
+            // against a store such as Azure Blob Storage that supports eTag conditions on save.
+            while (true)
```

Comment:
The infinite loop here is quite problematic. Let's say the Bot is in a high load condition. The pattern then is:
1. Grab Semaphore. 
2. Do long running stuff. 
3. Try to save. 
4. Fail. 
5. Release Semaphore. 
6. Goto 1.  

... and under high load this will "fail" every time as a different machine will do a write as the lock window is so long. The odds of your particular machine being able to load / save approach 0 as overall system load increases. I can easily envision 4 or 5 machines handling the same conversation and each continually failing as the others are mucking with state. 

Worse, there's no escape / trigger here, so it'll loop forever.